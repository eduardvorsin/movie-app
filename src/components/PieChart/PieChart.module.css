.mask::before,
.mask::after {
	content: '';
	inset: 0;
	border-radius: 50%;
	position: absolute;
	mask: var(--mask);
}

.mask::before {
	background: var(--track-gradient);
}

.mask::after {
	background: var(--gradient);
}