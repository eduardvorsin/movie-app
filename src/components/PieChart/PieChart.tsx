'use client';
import { CSSProperties } from 'react';
import classes from './PieChart.module.css';
import { useTheme } from '@/context/ThemeProvider/ThemeProvider';
import { GeneralProps } from '@/types/shared';

const variants = {
	primary: {
		dark: '#85B8FF',
		light: '#0055CC',
	},
	rating: {
		'low': {
			dark: '#FF9C8F',
			light: '#AE2A19',
		},
		'medium': {
			dark: '#F5CD47',
			light: '#946F00',
		},
		'high': {
			dark: '#7EE2B8',
			light: '#216E4E',
		},
	},
}

export type Props = {
	size: number,
	value: number,
	barColor?: Record<'light' | 'dark', string>,
	trackColor?: Record<'light' | 'dark', string>,
	thickness?: number,
	appearance?: 'primary' | 'rating',
} & GeneralProps;

export default function PieChart({
	size,
	value,
	className,
	barColor,
	trackColor,
	thickness = 5,
	testId,
	appearance = 'primary',
	...props
}: Props) {
	const theme = useTheme();
	const themeTrackColor = theme === 'dark' ? '#38414A' : '#DCDFE4';

	let ratingType: keyof typeof variants['rating'];
	if (value >= 0 && value <= 40) {
		ratingType = 'low';
	} else if (value >= 40 && value <= 70) {
		ratingType = 'medium';
	} else {
		ratingType = 'high';
	}

	const themeBarColor = appearance === 'primary' ? variants['primary'][theme] : variants['rating'][ratingType][theme];

	const styles = {
		'--width': `${size}px`,
		'--height': `${size}px`,
		'--gradient': `conic-gradient(${barColor?.[theme] ?? themeBarColor} ${value}%, #0000 0)`,
		'--track-gradient': `conic-gradient(${trackColor?.[theme] ?? themeTrackColor} 100%, #0000 0)`,
		'--mask': `radial-gradient(farthest-side, #0000 calc(99% - ${thickness}px),#000 calc(100% - ${thickness}px))`,
	} as CSSProperties;

	const chartClasses = [
		className,
		'bg-neutral-0 dark:bg-dark-neutral-200 w-[--width] h-[--height] inline-flex items-center justify-center rounded-[50%] relative font-medium dark:text-dark-neutral-900 transition-colors duration-150'
	].join(' ');

	return (
		<div
			style={styles}
			className={chartClasses}
			data-testid={testId}
			{...props}
		>
			<div
				className={`${classes.mask}`}
			>
				<span
					className='flex items-center justify-center overflow-hidden'
					style={{
						'width': `${size - thickness * 2 - 5}px`,
						'height': `${size - thickness * 2 - 5}px`,
					}}
				>
					{value}%
				</span>
			</div>
		</div>
	);
};
