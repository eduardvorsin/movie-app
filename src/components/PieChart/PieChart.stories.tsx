import type { Meta, StoryObj } from '@storybook/react';
import PieChart, { Props } from './PieChart';

const meta: Meta<typeof PieChart> = {
	title: 'UI/PieChart',
	tags: ['autodocs'],
	argTypes: {
		size: {
			description: 'Chart width and height',
		},
		className: {
			description: 'Additional classes for the component'
		},
		value: {
			description: 'The fullness value of the progress bar',
		},
		barColor: {
			description: 'Allows you to set the color of the progress bar',
		},
		testId: {
			description: 'Id for testing the component',
		},
		trackColor: {
			description: 'Defines the color of the pie chart track',
		},
		thickness: {
			description: 'Sets the thickness of the strips in the diagram',
		},
		appearance: {
			description: 'Defines the type of appearance of the component.',
		},
	},
	component: PieChart,
	parameters: {
		docs: {
			description: {
				component: 'The PieChart component provides a graphical representation of the data in the form of a pie chart (or pie)',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof PieChart>;

export const Default: Story = {
	args: {
		size: 60,
		value: 33,
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const BarColor: Story = {
	args: {
		value: 50,
		size: 60,
		barColor: {
			light: '#5600e8',
			dark: '#dbb2ff',
		},
	},
	parameters: {
		docs: {
			description: {
				story: 'Custom color for the progress line',
			},
		},
	},
};

export const TrackColor: Story = {
	args: {
		size: 60,
		value: 66,
		barColor: {
			light: '#dbb2ff',
			dark: '#5600e8',
		},
	},
	parameters: {
		docs: {
			description: {
				story: 'Custom color for the track',
			},
		},
	},
};

export const Thickness: Story = {
	args: {
		size: 60,
		value: 40,
		barColor: {
			light: '#dbb2ff',
			dark: '#5600e8',
		},
	},
	parameters: {
		docs: {
			description: {
				story: 'Custom line thickness for the chart',
			},
		},
	},
};

const appearances: NonNullable<Props['appearance']>[] = ['primary', 'rating'];
export const Appearance: Story = {
	parameters: {
		docs: {
			description: {
				story: 'The appearance property allows you to set color options for the component. PieChart can have 2 options: primary and rating. The second option, depending on the value, turns the progress bar red, yellow or green',
			},
		},
	},
	render: () => (
		<div className='flex items-start gap-3 flex-wrap'>
			{appearances.map((appearance) => (
				<PieChart
					key={appearance}
					value={75}
					size={60}
					appearance={appearance}
				/>
			))}
		</div>
	),
};
