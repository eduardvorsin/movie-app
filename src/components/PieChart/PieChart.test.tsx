import { render, screen } from "@testing-library/react";
import PieChart, { Props } from "./PieChart";

const appearances: NonNullable<Props['appearance']>[] = ['primary', 'rating'];

describe('PieChart tests', () => {
	it('is rendered correctly', () => {
		render(
			<PieChart
				value={50}
				size={50}
				testId='test-pie-chart'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-pie-chart')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<PieChart
				value={50}
				size={50}
				testId='test-pie-chart'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-pie-chart')).toMatchSnapshot();
	});

	it.each(appearances)('a snapshot with appearance equal to %s', (appearance) => {
		render(
			<PieChart
				value={50}
				size={50}
				appearance={appearance}
				testId='test-pie-chart'
			/>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-pie-chart')).toMatchSnapshot();
	});
});