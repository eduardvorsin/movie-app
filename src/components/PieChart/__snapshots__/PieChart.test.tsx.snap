// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChart tests a snapshot with appearance equal to primary 1`] = `
<div
  class=" bg-neutral-0 dark:bg-dark-neutral-200 w-[--width] h-[--height] inline-flex items-center justify-center rounded-[50%] relative font-medium dark:text-dark-neutral-900 transition-colors duration-150"
  data-testid="test-pie-chart"
  style="--width: 50px; --height: 50px; --gradient: conic-gradient(#0055CC 50%, #0000 0); --track-gradient: conic-gradient(#DCDFE4 100%, #0000 0); --mask: radial-gradient(farthest-side, #0000 calc(99% - 5px),#000 calc(100% - 5px));"
>
  <div
    class="before:absolute before:inset-0 before:rounded-full before:[background:var(--track-gradient)] before:[mask:var(--mask)] after:absolute after:inset-0 after:rounded-full after:[background:var(--gradient)] after:[mask:var(--mask)]"
  >
    <span
      class="flex items-center justify-center overflow-hidden"
      style="width: 35px; height: 35px;"
    >
      50
      %
    </span>
  </div>
</div>
`;

exports[`PieChart tests a snapshot with appearance equal to rating 1`] = `
<div
  class=" bg-neutral-0 dark:bg-dark-neutral-200 w-[--width] h-[--height] inline-flex items-center justify-center rounded-[50%] relative font-medium dark:text-dark-neutral-900 transition-colors duration-150"
  data-testid="test-pie-chart"
  style="--width: 50px; --height: 50px; --gradient: conic-gradient(#946F00 50%, #0000 0); --track-gradient: conic-gradient(#DCDFE4 100%, #0000 0); --mask: radial-gradient(farthest-side, #0000 calc(99% - 5px),#000 calc(100% - 5px));"
>
  <div
    class="before:absolute before:inset-0 before:rounded-full before:[background:var(--track-gradient)] before:[mask:var(--mask)] after:absolute after:inset-0 after:rounded-full after:[background:var(--gradient)] after:[mask:var(--mask)]"
  >
    <span
      class="flex items-center justify-center overflow-hidden"
      style="width: 35px; height: 35px;"
    >
      50
      %
    </span>
  </div>
</div>
`;

exports[`PieChart tests is a basic snapshot 1`] = `
<div
  class=" bg-neutral-0 dark:bg-dark-neutral-200 w-[--width] h-[--height] inline-flex items-center justify-center rounded-[50%] relative font-medium dark:text-dark-neutral-900 transition-colors duration-150"
  data-testid="test-pie-chart"
  style="--width: 50px; --height: 50px; --gradient: conic-gradient(#0055CC 50%, #0000 0); --track-gradient: conic-gradient(#DCDFE4 100%, #0000 0); --mask: radial-gradient(farthest-side, #0000 calc(99% - 5px),#000 calc(100% - 5px));"
>
  <div
    class="before:absolute before:inset-0 before:rounded-full before:[background:var(--track-gradient)] before:[mask:var(--mask)] after:absolute after:inset-0 after:rounded-full after:[background:var(--gradient)] after:[mask:var(--mask)]"
  >
    <span
      class="flex items-center justify-center overflow-hidden"
      style="width: 35px; height: 35px;"
    >
      50
      %
    </span>
  </div>
</div>
`;
